File: typeScript/challenges/challenge12.ts
interface OrnamentsValues {
  '*': number
  o: number
  '^': number
  '#': number
  '@': number
}

export function calculatePrice (ornaments: string): number | undefined {
  // Evaluate if there it is invalid chars
  const regex = /^[#@*o^]+$/
  if (!regex.test(ornaments)) return undefined

  // Ornaments values
  const values: OrnamentsValues = {
    '*': 1,
    o: 5,
    '^': 10,
    '#': 50,
    '@': 100
  }

  // Get the last value
  let price: number = values[ornaments.at(-1) as keyof OrnamentsValues]

  // Evaluate the ornamemnt i and the ornament i + 1
  for (let i = 0; i < ornaments.length - 1; i += 1) {
    const value1: number = values[ornaments.at(i) as keyof OrnamentsValues]
    const value2: number = values[ornaments.at(i + 1) as keyof OrnamentsValues]
    if (value2 > value1) {
      price -= value1
      continue
    }
    price += value1
  }
  return price
}

File: python/challenge12.py
def calculate_price(ornaments: str) -> int:
    ornaments_dict = {
        '*':1,
        'o':5,
        '^':10,
        '#':50,
        '@':100
    }
    values = []
    for i, j in enumerate(ornaments):
        if j not in ornaments_dict:
            return 'undefined'
        if i > 0:
            if ornaments_dict[j] > ornaments_dict[ornaments[i-1]]:
                values.append(ornaments_dict[ornaments[i-1]]*-1)
            else:
                values.append(ornaments_dict[ornaments[i-1]])
    values.append(ornaments_dict[ornaments[-1]])
    #print(values)
    return sum(values)

print(calculate_price('***'))
print(calculate_price('*o'))
print(calculate_price('*o*'))
print(calculate_price('*o@'))